### PR TITLE
fix(euclidean_cluster): add intensity field

### PR DESCRIPTION
**Pull request by badai-nguyen**
## Description

- To process and output the fields of clusters' pointcloud as the input topic including `intensity` field for `voxel_grid_based_euclidean_cluster` to solve the issue https://github.com/autowarefoundation/autoware.universe/issues/6785

## Related links

- [TIER IV INTERNAL LINK](https://tier4.atlassian.net/browse/RT1-5996?focusedCommentId=166391)

## Tests performed

- Confirmed by logging_simulator that voxel_grid_based_euclidean_cluster
   - Can process and output the topic with intensity field:

image

   - Processing time increase a little bit for a quite complicated scene
image

| Before | After | 
| -- | -- | 
|image| image |

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Comment by badai-nguyen** [comment]
@YoshiRi could I ask your review for this PR also?

File: perception/euclidean_cluster/lib/voxel_grid_based_euclidean_cluster.cpp [hunk 1]
```diff
@@ -40,14 +40,27 @@ VoxelGridBasedEuclideanCluster::VoxelGridBasedEuclideanCluster(
...
+  const sensor_msgs::msg::PointCloud2::ConstSharedPtr & pointcloud_msg,
+  tier4_perception_msgs::msg::DetectedObjectsWithFeature & objects)
 {
   // TODO(Saito) implement use_height is false version
 
   // create voxel
+  pcl::PointCloud<pcl::PointXYZ>::Ptr pointcloud(new pcl::PointCloud<pcl::PointXYZ>);
+  int point_step = pointcloud_msg->point_step;
+  pcl::fromROSMsg(*pointcloud_msg, *pointcloud);
   pcl::PointCloud<pcl::PointXYZ>::Ptr voxel_map_ptr(new pcl::PointCloud<pcl::PointXYZ>);
```
Comment (YoshiRi) [inline perception/euclidean_cluster/lib/voxel_grid_based_euclidean_cluster.cpp:64]: [minor comment]
For readability, I like following style.

```suggestion
  constexpr float z_leaf_size = 100000.0;
  voxel_grid_.setLeafSize(voxel_leaf_size_, voxel_leaf_size_, z_leaf_size);
```

**Review by YoshiRi — APPROVED** [review]
LGTM

**Comment by badai-nguyen** [comment]
@yukkysaito Do you have any comment for this PR?
Otherwise, could you approve this PR as codeowner?

**Review by yukkysaito — APPROVED** [review]
LGTM

File: perception/euclidean_cluster/src/voxel_grid_based_euclidean_cluster_node.cpp [hunk 2]
```diff
@@ -55,24 +55,15 @@ void VoxelGridBasedEuclideanClusterNode::onPointCloud(
...
     // NOTE: prevent pcl log spam
     RCLCPP_WARN_STREAM_THROTTLE(
       this->get_logger(), *this->get_clock(), 1000, "Empty sensor points!");
-  } else {
-    pcl::fromROSMsg(*input_msg, *raw_pointcloud_ptr);
   }
-
-  // clustering
-  std::vector<pcl::PointCloud<pcl::PointXYZ>> clusters;
-  if (!raw_pointcloud_ptr->empty()) {
```
Comment (firemark) [inline perception/euclidean_cluster/src/voxel_grid_based_euclidean_cluster_node.cpp:69]: I know, it's old pull request, but in carla simulator, logs are flooded from PCL errors about empty cloud. After reading `VoxelGridBasedEuclideanCluster::cluster` I think clustering empty cloud still produces empty output, so is safe to restore checking if cloud is empty?

```cpp
if (!input_msg->data.empty()) {
  cluster_->cluster(input_msg, output);
}
```

I can make a new PR but ain't sure it's a good idea.

**Pull request merged**